source/server/session_repository.cpp

```diff
@@ -65,6 +65,17 @@ uint32_t SessionRepository::access_session(uint32_t session_id, const std::strin
   return 0;
 }
 
+// Updates cleanup_func of a session by ID.
+// To update cleanup_func of a session by name, use access_session to get the session ID.
+void SessionRepository::update_cleanup_func(uint32_t id, CleanupSessionFunc new_cleanup_func)
+{
+  std::unique_lock<std::shared_mutex> lock(repository_lock_);
+  auto it = sessions_.find(id);
+  if (it != sessions_.end()) {
```

Comment:
There are three scenarios this can be called in: no session found, session already mapped to same close function, and session mapped to different close function.
- If the session ID doesn't exist, this method doesn't do anything.
- If the session exists and is already mapped to the same close function as the one passed in, this doesn't do anything useful.
- If the session ID was added from a call to the normal init function, then it's never correct to map it to a custom close function. The same applies the other direction. Imagine a user accidentally passing the driver session into a call to `CloseExtCal`.

For these reasons, you should remove this method.

@reckenro, @kylesull30, I'm afraid this is pointing at a bigger issue with our session management. I noticed in testing that drivers return incrementing I as session IDs. This means two drivers are likely to return the same session ID, but our repository doesn't keep IDs from different drivers separate from each other. As soon as you try using the server with multiple drivers you'll start getting errors.

Reply:
@jonlugoatx while calling the update clean up lambda we are not passing a different function but passing different parameters compared to what was registered during InitExtCal